frontend/class-frontend-page-type.php

```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * @package WPSEO\Frontend
+ */
+
+/**
+ * Represents the classifier for determine the current opened page type.
+ */
+class WPSEO_Frontend_Page_Type {
+
+	/**
+	 * @var WPSEO_WooCommerce_Shop_Page
+	 */
+	protected $woocommerce_shop_page;
+
+	/**
+	 * Sets the WooCommmerce shop page object.
+	 *
+	 * @param WPSEO_WooCommerce_Shop_Page $woocommerce_shop_page The woocommerce shop page object.
+	 */
+	public function __construct( WPSEO_WooCommerce_Shop_Page $woocommerce_shop_page ) {
+		$this->woocommerce_shop_page = $woocommerce_shop_page;
+	}
+
+	/**
+	 * Checks if the current opened page is singular.
+	 *
+	 * @return bool True when singular.
+	 */
+	public function is_singular() {
+		if ( is_singular() ) {
+			return true;
+		}
+
+		if ( is_home() && 'page' === get_option( 'show_on_front' ) ) {
+			return true;
+		}
+
+		if ( $this->woocommerce_shop_page->is_shop_page() ) {
```

Comment:
How do you see this growing in the future? Just a list of if statements for all the different cases? Is there a better way?

Reply:
My first thought was introducing a hook. Maybe that will be a better solution for this.